msm/src/serialization/constraints.rs

```diff
@@ -1,14 +1,88 @@
-use ark_ff::PrimeField;
+use ark_ff::{Field, PrimeField, Zero};
 use kimchi::circuits::{
-    expr::{ConstantExpr, ConstantTerm, Expr, ExprInner, Variable},
+    expr::{ChallengeTerm, ConstantExpr, ConstantTerm, Expr, ExprInner, Variable},
     gate::CurrOrNext,
 };
 
-use crate::{columns::Column, expr::E, serialization::N_INTERMEDIATE_LIMBS, N_LIMBS};
+use crate::{
+    columns::Column,
+    expr::{curr_cell, E},
+    serialization::N_INTERMEDIATE_LIMBS,
+    MVLookupTableID as _, N_LIMBS,
+};
 
 use super::Lookup;
 use super::{interpreter::InterpreterEnv, LookupTable};
 
+/// Compute the following constraint:
+/// ```text
+///                     lhs
+///    |------------------------------------------|
+///    |                           denominators   |
+///    |                         /--------------\ |
+/// column * (\prod_{i = 1}^{N} (\beta + f_{i}(X))) =
+/// \sum_{i = 1}^{N} m_{i} * \prod_{j = 1, j \neq i}^{N} (\beta + f_{j}(X))
+///    |             |--------------------------------------------------|
+///    |                             Inner part of rhs                  |
+///    |                                                                |
+///    |                                                               /
+///     \                                                             /
+///      \                                                           /
+///       \---------------------------------------------------------/
+///                           rhs
+/// ```
+pub fn combine_lookups<F: Field>(column: Column, lookups: Vec<Lookup<E<F>>>) -> E<F> {
+    let joint_combiner = {
+        let joint_combiner = ConstantExpr::from(ChallengeTerm::JointCombiner);
+        E::Atom(ExprInner::Constant(joint_combiner))
+    };
+    let beta = {
+        let beta = ConstantExpr::from(ChallengeTerm::Beta);
+        E::Atom(ExprInner::Constant(beta))
+    };
+
+    // Compute (\beta + f_{i}(X)) for each i.
+    // Note that f_i(X) = x_{0} + r x_{1} + ... r^{N} x_{N} + r^{N + 1} table_id
+    let denominators = lookups
+        .iter()
+        .map(|x| {
+            let combined_value = (x
+                .value
+                .iter()
+                .rev()
+                .fold(E::zero(), |acc, y| acc * joint_combiner.clone() + y.clone())
```

Comment:
Oh okay, so here the table_id goes with the largest power... I was wondering if it made sense to randomize all values. Meaning, start with the `r^1` power instead of `r^0`. Otherwise, in case table_id is 0 and the rest of values are zeros except for `x_{0}`, then `x_{0}` is not randomized. Just wondering if this could hide any small vulnerability.

Reply:
So perhaps one could do `r * table_id + r^2 * x_{0} + ....`